File: src/mp3.c
#include "mp3.h"

/**
 * Returns the audio quality depending on which mode we are
 * using and the quality parameter.
 */
int32_t mp3_get_output_quality(int32_t mode, int32_t quality) {
  if (mode == MP3_VBR_MODE) {
    return quality;
  } else {
    switch (quality) {
      default: return 320;
      case 0: return 32;
      case 1: return 48;
      case 2: return 64;
      case 3: return 96;
      case 4: return 128;
      case 5: return 160;
      case 6: return 192;
      case 7: return 224;
      case 8: return 256;
      case 9: return 320;
    }
  }
}

/**
 * Creates a structure used to keep lame and mp3_t data.
 */
EMSCRIPTEN_KEEPALIVE
mp3_t* mp3_create() {
  return malloc(sizeof(mp3_t));
}

/**
 * Frees all the used memory.
 */
EMSCRIPTEN_KEEPALIVE
int32_t mp3_destroy(mp3_t* e) {
  lame_close(e->gfp);
  free(e->input_buffer_left);
  free(e->input_buffer_right);
  free(e->output_buffer);
  free(e);
  return MP3_NO_ERROR;
}

/**
 * Initializes lame and all the buffers.
 */
EMSCRIPTEN_KEEPALIVE
int32_t mp3_init(mp3_t* e, int32_t sample_rate, int32_t num_channels, int32_t quality, int32_t samples, int32_t mode) {
  if (sample_rate < 11025 || sample_rate > 48000) {
    return MP3_ERROR_CONFIG_SAMPLE_RATE;
  }

  if (num_channels < 1 || num_channels > 2) {
    return MP3_ERROR_CONFIG_NUM_CHANNELS;
  }

  if (quality < 0 || quality > 9) {
    return MP3_ERROR_CONFIG_QUALITY;
  }

  if (samples < 2048 || samples > 16384) {
    return MP3_ERROR_CONFIG_SAMPLES;
  }

  // Inicializamos la estructura de LAME.
  e->gfp = lame_init();
  if (!e->gfp) {
    return MP3_ERROR_LAME_INIT;
  }

  // Reservamos la memoria necesaria para copiar las muestras que van
  // a entrar a través del micrófono.
  e->input_buffer_left = malloc(samples * sizeof(float));
  e->input_buffer_right = malloc(samples * sizeof(float));

  // Muestras de entrada.
  e->input_samples = samples;
  e->input_samplerate = sample_rate;
  e->input_channels = num_channels;

  e->output_samplerate = MP3_DEFAULT_OUTPUT_SAMPLE_RATE;
  e->output_quality = mp3_get_output_quality(mode, quality);

  // Reservamos 1MB para ir tirando. Aquí se irá almacenando
  // todo lo que necesitemos a la hora de codificar el MP3.
  e->output_buffer_max_size = samples * 1.25 + 7200;
  e->output_buffer_size = 0;
  e->output_buffer = malloc(e->output_buffer_max_size);

  // Establecemos la configuración de LAME.
  lame_set_in_samplerate(e->gfp, e->input_samplerate);
  lame_set_out_samplerate(e->gfp, e->output_samplerate);
  lame_set_num_channels(e->gfp, e->input_channels);
  if (mode == MP3_VBR_MODE) {
    lame_set_VBR(e->gfp, vbr_default);
    lame_set_VBR_q(e->gfp, e->output_quality);
  } else {
    lame_set_VBR(e->gfp, vbr_off);
    lame_set_brate(e->gfp, e->output_quality);
  }

  // Reservamos la memoria necesaria para los settings
  // que hemos establecido.
  if (lame_init_params(e->gfp) < 0) {
    return MP3_ERROR_LAME_INIT_PARAMS;
  }

  return MP3_NO_ERROR;
}

/**
 * Encodes MP3 frames.
 */
EMSCRIPTEN_KEEPALIVE
int32_t mp3_encode(mp3_t* e, int32_t num_samples) {
  // If 0 samples are passed we consider that we have already finished
  // to encode the mp3 so let's flush what we have in
  // the internal buffer.
  if (num_samples == 0) {
    return mp3_flush(e);
  }

  // If negative samples are passed this returns an error.
  if (num_samples < 0) {
    return MP3_ERROR_NUM_SAMPLES;
  }

  // If more samples than allowed are passed, then
  // we return an error.
  if (num_samples > e->input_samples) {
    return MP3_ERROR_TOO_MUCH_SAMPLES;
  }

  // Encodes the buffer and returns the number of bytes
  // that you have encoded.
  float* right = e->input_channels == 2 ? e->input_buffer_right : NULL;
  int32_t num_bytes = lame_encode_buffer_ieee_float(
    e->gfp,
    e->input_buffer_left,
    right,
    num_samples,
    e->output_buffer,
    e->output_buffer_max_size
  );

  if (num_bytes < 0) {
    return num_bytes - 100;
  }
  e->output_buffer_size = num_bytes;
  return MP3_NO_ERROR;
}

/**
 * Prevents losing half encoded frames.
 */
int32_t mp3_flush(mp3_t* e) {
  int32_t num_bytes = lame_encode_flush(e->gfp, e->output_buffer, e->output_buffer_max_size);
  if (num_bytes < 0) {
    return MP3_ERROR_LAME_ENCODE_FLUSH;
  }
  e->output_buffer_size = num_bytes;
  return MP3_NO_ERROR;
}

File: src/mp3.js
/**
 * Loads the `mp3.wasm` and returns the main class `Encoder`.
 * @module @etercast/mp3
 *
 * @func instantiate
 * @param {string|URL|Request} [wasm='mp3.wasm'] WASM file URL.
 * @returns Promise<class Encoder|Error>
 */
export function instantiate(wasm = './mp3.wasm') {
  return fetch(wasm)
    .then((response) => response.arrayBuffer())
    .then((arrayBuffer) => WebAssembly.instantiate(arrayBuffer, {
      // TODO: We really don't need these imports because mp3.wasm
      // does not handle file descriptors.
      wasi_unstable: {
        fd_close() { console.log('fd_close') },
        fd_seek() { console.log('fd_seek') },
        fd_write() { console.log('fd_write') },
        proc_exit() { console.log('proc_exit') }
      }
    }))
    .then(({ instance }) => {
      // TODO: There's an extra export called _start that's not used
      // but is defined in WASI. Look for ways to remove this unnecessary
      // export.
      const {
        memory,
        mp3_create,
        mp3_init,
        mp3_encode,
        mp3_destroy
      } = instance.exports

      /**
       * Returns the error codes from the MP3 encoder.
       * @param {number} errorCode
       * @return {string}
       */
      function getEncoderError(errorCode) {
        switch (errorCode) {
          default:
            return `Unknown error: ${errorCode}`
          case -1:
            return 'Invalid input sample rate'
          case -2:
            return 'Invalid input channels (mono and stereo only)'
          case -3:
            return 'Invalid quality'
          case -4:
            return 'Error calling lame_init'
          case -5:
            return 'Error calling lame_init_params'
          case -6:
            return 'Error reallocating buffers'
          case -7:
            return 'Too much input samples'
          case -8:
            return 'Error calling lame_encode_buffer_ieee_float'
          case -9:
            return 'Error calling lame_encode_flush'
          case -10:
            return 'Invalid number of samples passed'
          case -11:
            return 'Invalid input samples'
          case -100:
            return 'Ok'
          case -101:
            return 'Error calling lame_encode_buffer_ieee_float: Buffer was too small'
          case -102:
            return 'Error calling lame_encode_buffer_ieee_float: malloc() problem'
          case -103:
            return 'Error calling lame_encode_buffer_ieee_float: lame_init_params not called'
          case -104:
            return 'Error calling lame_encode_buffer_ieee_float: Psycho acoustic problems'
          case -110:
            return 'Error calling lame_encode_buffer_ieee_float: No memory'
          case -111:
            return 'Error calling lame_encode_buffer_ieee_float: Bad bitrate'
          case -112:
            return 'Error calling lame_encode_buffer_ieee_float: Bad samplefreq'
          case -113:
            return 'Error calling lame_encode_buffer_ieee_float: Internal error'
        }
      }

      /**
       * Encoder states.
       * @readonly
       * @enum {string}
       */
      const EncoderState = {
        /** Indicates that the encoder is running and is capable of encode MP3 frames. */
        RUNNING: 'running',
        /** Indicates that the encoder was closed and no longer can encode MP3 frames. */
        CLOSED: 'closed',
        /** Indicates that something went wrong. */
        ERROR: 'error'
      }

      /**
       * Encoder mode.
       * @readonly
       * @enum {number}
       */
      const EncoderMode = {
        /** Constant bit-rate */
        CBR: 0,
        /** Variable bit-rate */
        VBR: 1
      }

      /**
       * MP3 encoder options
       * @typedef {Object} EncoderOptions
       * @property {number} sampleRate Input/output sample rate. Usually this is the value from an AudioContext.
       * @property {number} numChannels Number of input/output channels. MP3 supports 1 (mono) or 2 (stereo) channels
       * @property {number} quality Encoding quality (0 - lowest, 9 - highest).
       *                              In VBR (Variable Bit-Rate) this quality indicates an average kbps but in
       *                              CBR (Constant Bit-Rate) 0 is equal to 32kbps and 9 is equal to 320kbps.
       * @property {number} samples Number of samples that will be encoded each time `encode` is called.
       * @property {EncoderMode} mode Encoding mode (0 - CBR, 1 - VBR).
       */

      /**
       * Not exported by default in the module, you need
       * to call `instantiate` to retrieve this class.
       *
       * @example
       * import instantiate from '@etercast/mp3'
       *
       * const Encoder = await instantiate()
       * const encoder = new Encoder(encoderOptions)
       * encoder.encode(leftChannelData, rightChannelData)
       * encoder.close()
       */
      class Encoder {
        /**
         * Creates a new MP3 encoder. This is equivalent to calling
         * the constructor using the `new` keyword. It's useful
         * if you need a function that instantiates the Encoder.
         * @param {EncoderOptions} options
         * @example
         * import instantiate from '@etercast/mp3'
         *
         * const Encoder = await instantiate()
         * const encoder = Encoder.create(encoderOptions)
         * @returns {Encoder}
         */
        static create(options) {
          return new Encoder(options)
        }

        /**
         * Internally this calls the exported method `mp3_create` to
         * make WASM module create a new structure to hold all the LAME
         * encoding data. It also calls `mp3_init` to establish encoder
         * options like number of channels, quality, sampleRate, etc.
         * @param {EncoderOptions} options
         */
        constructor(options) {
          const {
            sampleRate,
            numChannels,
            quality,
            samples,
            mode,
          } = {
            sampleRate: 44100,
            numChannels: 1,
            quality: 9,
            samples: 2048,
            mode: 0,
            ...options
          }
          this._error = null
          this._state = EncoderState.RUNNING
          this._pointer = null
          const pointer = mp3_create()
          if (!pointer) {
            return this._throw(new Error('Cannot create mp3 encoder'))
          }
          this._pointer = pointer
          const internalRegisters = 10
          const internal = new Int32Array(memory.buffer, pointer, internalRegisters)
          this._internal = internal
          const errorCode = mp3_init(this._pointer, sampleRate, numChannels, quality, samples, mode)
          if (errorCode < 0) {
            return this._throw(new Error(getEncoderError(errorCode)))
          }

          const [, outputBufferMaxSize, , , inputSamples, , , inputBufferLeftPointer, inputBufferRightPointer, outputBufferPointer] = internal
          this._inputBufferLeft = new Float32Array(memory.buffer, inputBufferLeftPointer, inputSamples)
          this._inputBufferRight = new Float32Array(memory.buffer, inputBufferRightPointer, inputSamples)
          this._outputBuffer = new Uint8Array(memory.buffer, outputBufferPointer, outputBufferMaxSize)
        }

        /**
         * Encoder state
         * @type {EncoderState}
         */
        get state() {
          return this._state
        }

        /**
         * Error
         * @type {null|Error}
         */
        get error() {
          return this._error
        }

        /**
         * Current output buffer size.
         * @type {number}
         */
        get outputBufferSize() {
          return this._internal[0]
        }

        /**
         * Max. output buffer size.
         * @type {number}
         */
        get outputBufferMaxSize() {
          return this._internal[1]
        }

        /**
         * Output sample rate
         * @type {number}
         */
        get outputSampleRate() {
          return this._internal[2]
        }

        /**
         * Output quality
         * @type {number}
         */
        get outputQuality() {
          return this._internal[3]
        }

        /**
         * Input samples
         * @type {number}
         */
        get inputSamples() {
          return this._internal[4]
        }

        /**
         * Input sample rate
         * @type {number}
         */
        get inputSampleRate() {
          return this._internal[5]
        }

        /**
         * Input number of channels (1 - mono, 2 - stereo)
         * @type {number}
         */
        get inputChannels() {
          return this._internal[6]
        }

        /**
         * Indicates that the encoder is running.
         * @type {boolean}
         */
        get isRunning() {
          return this._state === EncoderState.RUNNING
        }

        /**
         * Indicates that the encoder is running.
         * @type {boolean}
         */
        get isClosed() {
          return this._state === EncoderState.CLOSED
        }

        /**
         * Throws an error
         * @private
         * @param {Error} error
         */
        _throw(error) {
          this._state = EncoderState.ERROR
          this._error = error
          throw error
        }

        /**
         * Encodes raw Float 32-bit audio data into MP3 frames. An MP3
         * frame consist of a header and payload, this makes MP3 streamable
         * and implies that you can merge two files easily by concatenating them.
         *
         * If you need to merge two channels then you should use a `ChannelMergeNode`
         * and then reencode the audio.
         *
         * @example
         * import instantiate from '@etercast/mp3'
         *
         * const Encoder = await instantiate()
         * const encoder = new Encoder({
         *   sampleRate,
         *   samples,
         *   numChannels,
         *   quality,
         *   mode
         * })
         * encoder.encode(leftChannelData, rightChannelData)
         *
         * @param {Float32Array} left Left channel (mono)
         * @param {Float32Array} [right] Right channel (stereo)
         * @returns {Uint8Array} Returns a bunch of encoded frames
         */
        encode(left, right) {
          if (this._state !== EncoderState.RUNNING) {
            this._throw(new Error('Encoder already closed'))
          }
          let samples = 0
          if (left) {
            samples = left.length
            this._inputBufferLeft.set(left)
            if (right) {
              if (samples !== right.length) {
                this._throw(new Error('Encoder channels have different lengths'))
              }
              this._inputBufferRight.set(right)
            }
          }
          // Codifica el MP3.
          const errorCode = mp3_encode(this._pointer, samples)
          if (errorCode < 0) {
            this._throw(new Error(getEncoderError(errorCode)))
          }
          return this._outputBuffer.slice(0, this.outputBufferSize)
        }

        /**
         * Closes the encoder. After this method any call to `encode`
         * will throw an Error. If you need to append more data after
         * closing an encoder you can instantiate a new Encoder and then
         * concatenate all the new data with the old data.
         *
         * @example
         * import instantiate from '@etercast/mp3'
         *
         * const Encoder = await instantiate()
         * const encoder = new Encoder()
         *
         * // Do something with the encoder...
         *
         * encoder.close()
         *
         * // `encoder` is no longer usable and all the
         * // memory reserved is freed.
         */
        close() {
          if (this._state !== EncoderState.RUNNING) {
            this._throw(new Error('Encoder already closed'))
          }

          const errorCode = mp3_destroy(this._pointer)
          if (errorCode < 0) {
            this._throw(new Error(getEncoderError(errorCode)))
          }
          this._inputBufferLeft = null
          this._inputBufferRight = null
          this._outputBuffer = null
          this._pointer = null
          this._internal = null
          this._state = EncoderState.CLOSED
        }
      }

      return Encoder
    })
}

export default instantiate
